app/demo-geant-integration/HepMC3Reader.cc

```diff
@@ -0,0 +1,136 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HepMC3Reader.cc
+//---------------------------------------------------------------------------//
+#include "HepMC3Reader.hh"
+
+#include <mutex>
+#include <G4SystemOfUnits.hh>
+#include <HepMC3/ReaderFactory.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+
+#include "GlobalSetup.hh"
+
+namespace demo_geant
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Return non-owning pointer to a singleton.
+ */
+HepMC3Reader* HepMC3Reader::instance()
+{
+    static HepMC3Reader hepmc3_reader_singleton;
+    return &hepmc3_reader_singleton;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add HepMC3 primaries to a Geant4 event. This function is called by
+ * `G4VUserPrimaryGeneratorAction::GeneratePrimaries`.
+ */
+void HepMC3Reader::GeneratePrimaryVertex(G4Event* g4_event)
+{
+    static std::mutex           hepmc3_mutex;
+    std::lock_guard<std::mutex> scoped_lock{hepmc3_mutex};
+
+    const auto primaries = this->load_primaries();
```

Comment:
Is there a reason we need this helper function/extra `Primary` struct, or could the reading be done inside this method (so we're only looping over the primaries once)?